# Lab Assignment 8: Data Management Using `pandas`, Part 1
## DS 6001

### Instructions
Please answer the following questions as completely as possible using text, code, and the results of code as needed. Format your answers in a Jupyter notebook. To receive full credit, make sure you address every part of the problem, and make sure your document is formatted in a clean and professional way.


## Problem 0
Import the following packages:

In [271]:
import numpy as np
import pandas as pd
import sidetable
import sqlite3

## Problem 1
There are a lot of resources on the internet on the general topic of Python for Data Science, and many of these websites publish lists of the most important Python packages for data science. There are definite disagreements further down on these lists, but the [top](https://www.datacamp.com/blog/top-python-libraries-for-data-science) [spots](https://www.reddit.com/r/Python/comments/uorxt5/does_anyone_use_python_for_a_data_science/) [always](https://www.anaconda.com/topics/best-python-libraries) [include](https://www.geeksforgeeks.org/python-libraries-for-data-science/) numpy and pandas. We've used pandas in every single one of the labs in this course so far, and modules 8 and 9 are entirely about the process of data wrangling and manipulation using pandas.

But, there is a package that competes directly with pandas, called [polars](https://pola.rs/). This package has its fans and adherents. But the tenor and tone of the debate over which package to use is more adversarial and heated than I would expect. For many, it is not enough to use polars when it makes sense as part of a larger data science toolkit: polars needs to completely defeat and replace pandas, and send pandas to that great software recycling bin in the sky.

As a data professional, you will absolutely be drawn into a conversation at some point that's in reality a small battle in the great pandas vs. polars war. Maybe polars really is the inevitable replacement for pandas, maybe not. But you should know about these options to form a solid opinion so that you can hold your ground in one of these conversations until civility can be reestablished and everyone can just talk about sports or gossip again.

### Part a
The polars website contains [a page that is directed at pandas users](https://docs.pola.rs/user-guide/migration/pandas/). Please read this page, and find two points made on this page that seem appealing or interesting to you, and describe those points here. [4 points]

<strong> I find it interesting that polars does not require the use of as many lambda functions, instead allowing for more expressions. I'm curious how this works and if it is easier in practice.

Additionally, I like that polars generally appears to use methods, instead of using `.loc`/`.iloc`. I don't know enough to comment yet on if this is better, but my instincts find it to be a safer, more consistent, more understandable. I read *Clean Code* by Robert C. Martin a long time ago, and it impressed upon me the necessity for code to be clear and readable. Polars, in this aspect, appears to be so.</strong>

### Part b
So, here's the thing. When there's contention between a new and an old way of doing something, the voices on the side of the new way will usually be louder, though not necessarily greater in number. "This New Way Will Change Everything" makes a more clickable blog title than "There's Nothing Wrong With the Way We Do Things Now." To see the argument for pandas over polars, look for analyses that aim for an objective comparison. Read this [blog post by Josep Ferrer](https://www.kdnuggets.com/pandas-vs-polars-a-comparative-analysis-of-python-dataframe-libraries), or find a different post or analysis that defends pandas, and in a one or two of your own sentences, describe the best argument for pandas. [4 points]

<strong>Pandas is the better library for machine learning and data visualization because of its compatibility with the main python libraries used for htese tasks. </strong>

### Part c
Do a search on a social media platform of your choice, and find an instance of somebody commenting on the polars vs. pandas debate. Copy and paste some text from this post that your find compelling or interesting, along with the link. Also, do you agree with this opinion, or not, and why? [4 points]

<strong>First, I just thought this was funny because I am constantly adding underscores in pandas, where they don't exist! I would tend to agree that, especially in the age of autocomplete and AI where we don't have to actually type out much of our code anymore, readability should be a top concern. 
>Polars separates words by underscores. "Group by" is two words, contrary to what Pandas would have you believe  

I feel a bit boring for choosing the [top voted reddit reply on the top google search](https://www.reddit.com/r/Python/comments/1jg402b/polars_vs_pandas/), but after reading around a bit, this is still what struck me the most:
>Feels like there's only one right way to do it in Polars, but so much bloat in Pandas API.

This resonates because I have been noticing as we learn more about pandas that there are a ton of different ways to do things. Having a ton of ways to approach a problem can be useful when a lot of flexibility is required, but it can also mean a high level of inefficiency. Not to mention the commenters concern of "bloat". For a new user, or even an experienced one from what I can tell in the comments, figuring out the "right" or "best" way to do something in pandas is a bit of a chore. And given it's exclusive use of lazy operators, doing something the best way is probably pretty important when it coems to speed and memory.</strong>

### Part d
As someone who either currently, or who will soon be a data professional, you will be a member of a community of similar professionals. We learn a great deal from our peers in these kinds of communities of practice. Have a conversation with someone who is a data professional, whether that is someone you work with, or someone you know, or someone that someone you know knows, and ask them about their thoughts on polars vs pandas. What are their thoughts on this? [4 points]


<strong>
I spoke to a friend of mine who is also studying Data Science and has done some research work. He said he prefers working with polars for his research work because of the size of the datasets they deal with, but that he appreciated learning pandas first. He was mainly glad he knew both because it gives him more flexibility.
</strong>

### Part e
Based on your reading and thoughts from parts (a) through (d), do you think you will try to learn how to use polars, or will you stick with pandas, or try using both? (No wrong answers, just curious about where you land on this.)

By the way: We are going to continue to use pandas in this course, but if you would like to train on polars, one good option is [this course](https://app.datacamp.com/learn/courses/introduction-to-polars) on Datacamp. You should have free access to Datacamp from UVA, but if not, please let me know so we can get you access. [4 points]

<strong>From the reading, I'm at least convinced that as a data professional I should try using it. It's a big part of the field right now, so deciding to forgo seems like a pretty limiting path. Additionally, it sounds interesting and like there are good arguments for its use!</strong>

## Problem 2

Please use pandas, not polars, for this problem.

For this problem, you will be working with the [2017 Workplace Health in America survey](https://www.cdc.gov/workplace-health-promotion/php/data/index.html) which was conducted by the Centers for Disease Control and Prevention. According to the survey's [guidance document](https://www.cdc.gov/workplace-health-promotion/media/pdfs/2024/06/2017-WHA-Guidance-Document-for-Use-of-Public-Data-files-508.pdf):

> The Workplace Health in America (WHA) Survey gathered information from a cross-sectional, nationally representative sample of US worksites. The sample was drawn from the Dun & Bradstreet (D&B) database of all private and public employers in the United States with at least 10 employees. Like previous national surveys, the worksite served as the sampling unit rather than the companies or firms to which the worksites belonged. Worksites were selected using a stratified simple random sample (SRS) design, where the primary strata were ten multi-state regions defined by the Centers for Disease Control and Prevention (CDC), plus an additional stratum containing all hospital worksites. 

The data contain over 300 features that report the industry and type of company where the respondents are employed, what kind of health insurance and other health programs are offered, and other characteristics of the workplaces including whether employees are allowed to work from home and the gender and age makeup of the workforce. The data are full of interesting information, but in order to make use of the data a great deal of data manipulation is required first.

### Part a
The raw data are stored in an ASCII file on the 2017 Workplace Health in America survey [homepage](https://www.cdc.gov/workplacehealthpromotion/survey/data.html). Load the raw data directly into Python without downloading the data onto your harddrive. Then use the `.info(verbose=True)` method on the dataframe to show summary information about the data. Based on what you see (it's NOT necessary to usE any other code) answer the following questions:

* How many rows?
* How many columns?
* How many columns have a float data type (numeric with decimals) and how many have an object data type (text or strings)?

[4 points]

In [ ]:
datalink = "https://www.cdc.gov/workplace-health-promotion/media/files/2024/06/whpps_120717.csv"
whads = pd.read_csv(datalink, delimiter="~")
whads.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2843 entries, 0 to 2842
Data columns (total 301 columns):
 #    Column                Dtype  
---   ------                -----  
 0    OC1                   object 
 1    OC3                   float64
 2    HI1                   float64
 3    HI2                   float64
 4    HI3                   float64
 5    HI4                   float64
 6    HRA1                  float64
 7    HRA1A                 float64
 8    HRA1B                 float64
 9    HRA1E                 float64
 10   CP1                   float64
 11   CP2                   float64
 12   CP3                   float64
 13   CP4                   float64
 14   CP5                   float64
 15   HP1                   float64
 16   HP2                   float64
 17   HP3                   float64
 18   HP4                   float64
 19   HP5                   float64
 20   HP5A                  float64
 21   HP6                   float64
 22   HP7A                  

/var/folders/fm/2_zxw4w57231pbsnw1dv90yc0000gn/T/ipykernel_63000/3947047464.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  whads = pd.read_csv(datalink, delimiter="~")


<strong>

* How many rows? 2843 rows
* How many columns? 301 columns
* How many columns have a float data type (numeric with decimals) and how many have an object data type (text or strings)? 
    7 have an `object` type and 294 have a `float64` type.

</strong>

### Part b 
Create a new variable in Python's memory to store a working version of the data. In the working version, delete all of the columns except for the following:

* `Industry`: 7 Industry Categories with NAICS codes

* `Size`: 8 Employee Size Categories

* `OC3` Is your organization for profit, non-profit, government?

* `HI1` In general, do you offer full, partial or no payment of premiums for personal health insurance for full-time employees?

* `HI2` Over the past 12 months, were full-time employees asked to pay a larger proportion, smaller proportion or the same proportion of personal health insurance premiums?

* `HI3`: Does your organization offer personal health insurance for your part-time employees?

* `CP1`: Are there health education programs, which focus on skill development and lifestyle behavior change along with information dissemination and awareness building?

* `WL6`: Allow employees to work from home?

* Every column that begins `WD`, expressing the percentage of employees that have certain characteristics at the firm

[4 points]

In [273]:
working_data = whads
cols = ['Industry', 'Size', 'OC3', 'HI1', 'HI2', 'HI3', 'CP1', 'WL6'] + [c for c in working_data.columns if c.startswith('WD')]
working_data = working_data[cols]
working_data.head()
len(working_data.columns)

16

### Part c
Column names like 'OC3' and 'CP1' do not convey the meaning or intuition of a feature, so let's choose better names. The [codebook](https://www.cdc.gov/workplace-health-promotion/media/pdfs/2024/06/2017-WHA-Datafile-Codebook-508.pdf) for the WHA data contain short descriptions of the meaning of each of the columns in the data. Use these descriptions to decide on better and more intuitive names for the columns in the working version of the data, and rename the columns accordingly. [4 points]

In [274]:
col_names = {
    'OC3': 'Org_Category',
    'HI1': 'Employer_Payment_Offer',
    'HI2': 'Proportion_Employee_Payment_Change',
    'HI3': 'Part_Time_Health_Insurance',
    'CP1': 'Health_Education_Programs',
    'WL6': 'Work_From_Home',
    'WD1_1': 'Percent_Age_LT_30',
    'WD1_2': 'Percent_Age_GTE_60',
    'WD2': 'Percent_Female',
    'WD3': 'Percent_Hourly_NonExempt',
    'WD4': 'Percent_NonDaytime_Shift',
    'WD5': 'Percent_Remote',
    'WD6': 'Percent_Unionized',
    'WD7': 'Percent_Annual_Turnover'
}
working_data = working_data.rename(col_names, axis=1)
working_data

,Industry,Size,Org_Category,Employer_Payment_Offer,Proportion_Employee_Payment_Change,Part_Time_Health_Insurance,Health_Education_Programs,Work_From_Home,Percent_Age_LT_30,Percent_Age_GTE_60,Percent_Female,Percent_Hourly_NonExempt,Percent_NonDaytime_Shift,Percent_Remote,Percent_Unionized,Percent_Annual_Turnover
0,7.0,7.0,3.0,2.0,1.0,2.0,1.0,1.0,25.0,20.0,85.0,60.0,40.0,15.0,0.0,22.0
1,7.0,6.0,3.0,2.0,3.0,1.0,1.0,1.0,997.0,997.0,90.0,90.0,997.0,997.0,0.0,997.0
2,7.0,8.0,3.0,1.0,3.0,1.0,1.0,1.0,35.0,4.0,997.0,997.0,40.0,15.0,997.0,997.0
3,7.0,4.0,2.0,1.0,2.0,1.0,2.0,2.0,50.0,15.0,50.0,85.0,75.0,0.0,0.0,997.0
4,7.0,4.0,3.0,1.0,3.0,1.0,1.0,1.0,50.0,40.0,60.0,60.0,40.0,30.0,0.0,28.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2838,6.0,5.0,4.0,1.0,3.0,1.0,1.0,99.0,999.0,999.0,999.0,999.0,999.0,999.0,999.0,999.0
2839,6.0,5.0,4.0,2.0,3.0,1.0,1.0,2.0,997.0,997.0,997.0,997.0,997.0,997.0,997.0,997.0
2840,6.0,8.0,4.0,2.0,3.0,1.0,1.0,1.0,27.0,997.0,61.0,997.0,997.0,997.0,997.0,997.0
2841,6.0,8.0,4.0,2.0,3.0,1.0,2.0,99.0,999.0,999.0,999.0,999.0,999.0,999.0,999.0,999.0


### Part d
Throughout the data, numeric values are used to represent categories. For example, for the feature that asks whether a workplace allows employees to work from home, 1 means yes and 2 means no. But it is more intuitive, less error prone, and easier for data visualization and other analyses if we replace these numbers with text labels that tell us explicitly what the categories represent. The following questions ask you to replace numeric values with text labels for these categorical features.

#### Step i
The industry column is currently coded numerically:

In [275]:
working_data['Industry'].value_counts()

Industry
5.0    551
1.0    525
3.0    433
4.0    429
7.0    338
2.0    311
6.0    255
Name: count, dtype: int64

According to the codebook, these values represent:

* 1 = Industry Category 1: NAICS Sectors: 11, 21, 22, 23, 31-33
* 2 = Industry Category 2: NAICS Sectors: 42, 44-45, 48-49
* 3 = Industry Category 3: NAICS Sectors: 71, 72, 81
* 4 = Industry Category 4: NAICS Sectors: 51, 52, 53, 54, 55, 56
* 5 = Industry Category 5: NAICS Sectors: 61, 62 (excluding hospital worksites)
* 6 = Industry Category 6: NAICS Sectors: 92
* 7 = Industry Category 7: Hospital worksites (NAICS6 = 622110, 622210, 622310)

NAICS stands for North American Industry Classification System. According to their [website](https://www.naics.com/), their industry coding system

> is a standardized system used to classify businesses based on their primary activities. This NAICS Code classification helps in the collection, analysis, and publication of statistical data related to the U.S. economy. Accurate NAICS Information is crucial for businesses because they affect eligibility for government contracts, loans, and tax benefits, and they facilitate better communication with government agencies.

Use the [dictionary of NAICS industrial codes](https://www.naics.com/search-naics-codes-by-industry/) to lookup the NAICS codes that each category of the Industry feature represents, then place descriptive labels on the categories of the industry column in the working data. This is a bit of a pain because you need to find a succinct but descriptive way to label each category. For example, Industry=1 represents Agriculture, Forestry, Fishing and Hunting (NAICS code 11), Mining (NAICS code 21), Utilities (NAICS code 22), Construction (NAICS code 23), and Manufacturing (NAICS code 31-33). You can use the label 'Agriculture, Construction, Manufacturing' if you want, or something else. Do your best.

[4 points]

In [276]:
# 1: 	Agriculture, Forestry, Fishing and Hunting, Mining, Utilities, construction, and manufacturing
# 2:    Wholesale trade, retail trade, transportation, and warehouseing
# 3:    Arts, Entertainment, and Recreation, Accommodation and Food Services, 	Other Services (except Public Administration)
# 4:    Information, Finance and insurance, real estate rental and leasing, Professional, Scientific, and Technical Services,Management of Companies and Enterprises, Administrative and Support and Waste… Services
# 5:    Educational Services, Health Care and Social Assistance (excl. hospital worksites)
# 6:    Public Administration
# 7:    Hostpital Worksites
industry_categories = {
    1.0:'Food and Resource Production',
    2.0:'Trade and Transportation',
    3.0:'Entertainment and Non-public Services',
    4.0:'Info, Business, Tech, and Property',
    5.0:'Education, Health, and Social Assistance',
    6.0:'Public Administration',
    7.0:'Hospital Worksites'
}

In [277]:
working_data.Industry = working_data.Industry.map(industry_categories)
working_data[['Industry', 'Size']]

,Industry,Size
0,Hospital Worksites,7.0
1,Hospital Worksites,6.0
2,Hospital Worksites,8.0
3,Hospital Worksites,4.0
4,Hospital Worksites,4.0
...,...,...
2838,Public Administration,5.0
2839,Public Administration,5.0
2840,Public Administration,8.0
2841,Public Administration,8.0


#### Step ii
Using the codebook, recode the "size" feature to have three categories: "Small" for workplaces with fewer than 100 employees, "Medium" for workplaces with at least 100 but fewer than 500 employees, and "Large" for companies with at least 500 employees. [Note: Python dataframes have an attribute `.size` that reports the space the dataframe takes up in memory. Don't confuse this attribute with the column named "Size" in the raw data.] [4 points]

In [278]:
size_categories = {
    1.0:'Small',
    2.0:'Small',
    3.0:'Small',
    4.0:'Medium',
    5.0:'Medium',
    6.0:'Large',
    7.0:'Large',
    8.0:'Large'
}
working_data.Size = working_data.Size.map(size_categories)
working_data[['Industry', 'Size']]


,Industry,Size
0,Hospital Worksites,Large
1,Hospital Worksites,Large
2,Hospital Worksites,Large
3,Hospital Worksites,Medium
4,Hospital Worksites,Medium
...,...,...
2838,Public Administration,Medium
2839,Public Administration,Medium
2840,Public Administration,Large
2841,Public Administration,Large


#### Step iii
Use the codebook to write accurate and descriptive labels for each category for each categorical column in the working data. Then apply all of these labels to the data at once. Code "Legitimate Skip", "Don't know", "Refused", and "Blank" as `np.nan` missing values. [8 points]

In [279]:
working_data.columns

Index(['Industry', 'Size', 'Org_Category', 'Employer_Payment_Offer',
       'Proportion_Employee_Payment_Change', 'Part_Time_Health_Insurance',
       'Health_Education_Programs', 'Work_From_Home', 'Percent_Age_LT_30',
       'Percent_Age_GTE_60', 'Percent_Female', 'Percent_Hourly_NonExempt',
       'Percent_NonDaytime_Shift', 'Percent_Remote', 'Percent_Unionized',
       'Percent_Annual_Turnover'],
      dtype='object')

In [280]:
org_categories = {
    1.0:'For profit, public',
    2.0:'For profit, private',
    3.0:'Non-profit',
    4.0:'State or local government',
    5.0:'Federal government',
    6.0: 'Other',
    97.0: np.nan,
    98.0: np.nan,
    99.0: np.nan
}

employer_payment_categories = {
    1.0:'Full coverage',
    2.0:'Partial coverage',
    3.0:'No coverage',
    97.0: np.nan,
    98.0: np.nan,
    99.0: np.nan
}

employee_portion_change_categories = {
    1.0:'Larger',
    2.0:'Smaller',
    3.0:'About the same',
    96.0: np.nan,
    97.0: np.nan,
    98.0: np.nan,
    99.0: np.nan
}

part_time_insurance_categories = {
    1.0:'Offered',
    2.0:'Not offered',
    97.0: np.nan,
    98.0: np.nan,
    99.0: np.nan
}

health_education_categories = {
    1.0:'Offered',
    2.0:'Not offered',
    97.0: np.nan,
    98.0: np.nan
}

work_from_home_categories = {
    1.0:'Allowed',
    2.0:'Not allowed',
    97.0: np.nan,
    98.0: np.nan,
    99.0: np.nan
}

replace_map = {
    'Org_Category': org_categories, 
    'Employer_Payment_Offer': employer_payment_categories,
    'Proportion_Employee_Payment_Change': employee_portion_change_categories, 
    'Part_Time_Health_Insurance': part_time_insurance_categories,
    'Health_Education_Programs': health_education_categories,
    'Work_From_Home': work_from_home_categories
} 

In [281]:
working_data = working_data.replace(replace_map)
working_data.head()

,Industry,Size,Org_Category,Employer_Payment_Offer,Proportion_Employee_Payment_Change,Part_Time_Health_Insurance,Health_Education_Programs,Work_From_Home,Percent_Age_LT_30,Percent_Age_GTE_60,Percent_Female,Percent_Hourly_NonExempt,Percent_NonDaytime_Shift,Percent_Remote,Percent_Unionized,Percent_Annual_Turnover
0,Hospital Worksites,Large,Non-profit,Partial coverage,Larger,Not offered,Offered,Allowed,25.0,20.0,85.0,60.0,40.0,15.0,0.0,22.0
1,Hospital Worksites,Large,Non-profit,Partial coverage,About the same,Offered,Offered,Allowed,997.0,997.0,90.0,90.0,997.0,997.0,0.0,997.0
2,Hospital Worksites,Large,Non-profit,Full coverage,About the same,Offered,Offered,Allowed,35.0,4.0,997.0,997.0,40.0,15.0,997.0,997.0
3,Hospital Worksites,Medium,"For profit, private",Full coverage,Smaller,Offered,Not offered,Not allowed,50.0,15.0,50.0,85.0,75.0,0.0,0.0,997.0
4,Hospital Worksites,Medium,Non-profit,Full coverage,About the same,Offered,Offered,Allowed,50.0,40.0,60.0,60.0,40.0,30.0,0.0,28.0


#### Step iv
The features that measure the percent of the workforce with a particular characteristic use the codes 997, 998, and 999 to represent "Don't know", "Refusal", and "Blank/Invalid" respectively. If we don't change these, all numeric results that use these features will be wrong. Replace these values with missing values for all of the percentage features at the same time. [4 points]

In [282]:
nan_coding = {
    997: np.nan,
    998: np.nan,
    999: np.nan
}

percent_cols = [c for c in working_data.columns if c.startswith('Percent')]
working_data[percent_cols] = working_data[percent_cols].replace([997, 998, 999], np.nan)
working_data.head(10)

,Industry,Size,Org_Category,Employer_Payment_Offer,Proportion_Employee_Payment_Change,Part_Time_Health_Insurance,Health_Education_Programs,Work_From_Home,Percent_Age_LT_30,Percent_Age_GTE_60,Percent_Female,Percent_Hourly_NonExempt,Percent_NonDaytime_Shift,Percent_Remote,Percent_Unionized,Percent_Annual_Turnover
0,Hospital Worksites,Large,Non-profit,Partial coverage,Larger,Not offered,Offered,Allowed,25.0,20.0,85.0,60.0,40.0,15.0,0.0,22.0
1,Hospital Worksites,Large,Non-profit,Partial coverage,About the same,Offered,Offered,Allowed,NaN,NaN,90.0,90.0,NaN,NaN,0.0,NaN
2,Hospital Worksites,Large,Non-profit,Full coverage,About the same,Offered,Offered,Allowed,35.0,4.0,NaN,NaN,40.0,15.0,NaN,NaN
3,Hospital Worksites,Medium,"For profit, private",Full coverage,Smaller,Offered,Not offered,Not allowed,50.0,15.0,50.0,85.0,75.0,0.0,0.0,NaN
4,Hospital Worksites,Medium,Non-profit,Full coverage,About the same,Offered,Offered,Allowed,50.0,40.0,60.0,60.0,40.0,30.0,0.0,28.0
5,Hospital Worksites,Small,Non-profit,Full coverage,About the same,Offered,Offered,Allowed,20.0,25.0,65.0,80.0,25.0,5.0,0.0,15.0
6,Hospital Worksites,Large,Non-profit,Partial coverage,Larger,Offered,Offered,Allowed,20.0,20.0,80.0,70.0,60.0,5.0,10.0,NaN
7,Hospital Worksites,Large,"For profit, public",Partial coverage,About the same,Not offered,Offered,Not allowed,27.0,14.0,75.0,60.0,50.0,5.0,0.0,15.0
8,Hospital Worksites,Large,NaN,Full coverage,Larger,Offered,Offered,Allowed,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN
9,Hospital Worksites,Large,Non-profit,Partial coverage,About the same,Offered,Offered,Allowed,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Part e
As you proceed through a data wrangling process, viewing the data at each step is important to determine whether your code has worked correctly and to see the next problems that need to be solved. Sorting the data is one of the most important and useful ways to look at different parts of the data that are not displayed by default within the `.head()` of a dataframe.

Sort the working data by industry in ascending alphabetical order. Within industry categories, sort the rows by size in ascending alphabetical order. Within groups with the same industry and size, sort by percent of the workforce that is under 30 in descending numeric order. [4 points]

In [283]:
working_data.sort_values(by=['Industry', 'Size', 'Percent_Age_LT_30'], ascending=[True, True, False]).head(10)

,Industry,Size,Org_Category,Employer_Payment_Offer,Proportion_Employee_Payment_Change,Part_Time_Health_Insurance,Health_Education_Programs,Work_From_Home,Percent_Age_LT_30,Percent_Age_GTE_60,Percent_Female,Percent_Hourly_NonExempt,Percent_NonDaytime_Shift,Percent_Remote,Percent_Unionized,Percent_Annual_Turnover
1987,"Education, Health, and Social Assistance",Large,Non-profit,Partial coverage,About the same,Not offered,Offered,Allowed,46.0,10.0,86.0,75.0,50.0,50.0,0.0,35.0
1988,"Education, Health, and Social Assistance",Large,Non-profit,Full coverage,About the same,Offered,Offered,Allowed,45.0,25.0,65.0,60.0,60.0,5.0,NaN,7.0
1120,"Education, Health, and Social Assistance",Large,State or local government,Full coverage,About the same,Not offered,Offered,Not allowed,40.0,20.0,75.0,30.0,5.0,0.0,70.0,NaN
1641,"Education, Health, and Social Assistance",Large,Non-profit,Partial coverage,About the same,Offered,Not offered,Not allowed,40.0,40.0,60.0,40.0,NaN,NaN,50.0,NaN
2807,"Education, Health, and Social Assistance",Large,Non-profit,Full coverage,Smaller,Offered,Not offered,Not allowed,40.0,20.0,65.0,60.0,100.0,5.0,0.0,25.0
2806,"Education, Health, and Social Assistance",Large,State or local government,Full coverage,Smaller,Offered,Not offered,Not allowed,35.0,55.0,52.0,85.0,5.0,0.0,85.0,NaN
660,"Education, Health, and Social Assistance",Large,Non-profit,Partial coverage,Larger,Not offered,Offered,Not allowed,30.0,10.0,70.0,NaN,25.0,10.0,0.0,15.0
861,"Education, Health, and Social Assistance",Large,State or local government,Full coverage,About the same,Not offered,Offered,Allowed,30.0,5.0,45.0,73.0,2.0,98.0,0.0,10.0
1989,"Education, Health, and Social Assistance",Large,State or local government,Partial coverage,About the same,Offered,Offered,NaN,30.0,15.0,78.0,NaN,10.0,1.0,60.0,NaN
866,"Education, Health, and Social Assistance",Large,"For profit, public",Full coverage,Larger,Not offered,Offered,Allowed,25.0,20.0,55.0,NaN,15.0,NaN,0.0,NaN


### Part f
Sometimes important features have missing values for one reason or another, and depending on our purposes, we may not be able to use rows with these missing values. In those situations, it is often easier to delete the rows with missing values.

There is one row in the working data that has a missing value for industry. Delete this row.  [4 points]

In [284]:
working_data =working_data.dropna(subset=['Industry'])

### Part g
Sometimes we need to take a continuous-valued feature and turn it into a categorical one for some analysis.

Create a new feature named `gender_balance` that has three categories: "Mostly men" for workplaces with between 0% and 35% female employees, "Balanced" for workplaces with more than 35% and at most 65% female employees, and "Mostly women" for workplaces with more than 65% female employees. [4 points]

In [285]:
working_data['Gender_Balance'] = pd.cut(working_data['Percent_Female'], bins=[-0.1, 35 , 65, 100], labels=['Mostly men', 'Balanced', 'Mostly women'])
working_data.head(10)

,Industry,Size,Org_Category,Employer_Payment_Offer,Proportion_Employee_Payment_Change,Part_Time_Health_Insurance,Health_Education_Programs,Work_From_Home,Percent_Age_LT_30,Percent_Age_GTE_60,Percent_Female,Percent_Hourly_NonExempt,Percent_NonDaytime_Shift,Percent_Remote,Percent_Unionized,Percent_Annual_Turnover,Gender_Balance
0,Hospital Worksites,Large,Non-profit,Partial coverage,Larger,Not offered,Offered,Allowed,25.0,20.0,85.0,60.0,40.0,15.0,0.0,22.0,Mostly women
1,Hospital Worksites,Large,Non-profit,Partial coverage,About the same,Offered,Offered,Allowed,NaN,NaN,90.0,90.0,NaN,NaN,0.0,NaN,Mostly women
2,Hospital Worksites,Large,Non-profit,Full coverage,About the same,Offered,Offered,Allowed,35.0,4.0,NaN,NaN,40.0,15.0,NaN,NaN,NaN
3,Hospital Worksites,Medium,"For profit, private",Full coverage,Smaller,Offered,Not offered,Not allowed,50.0,15.0,50.0,85.0,75.0,0.0,0.0,NaN,Balanced
4,Hospital Worksites,Medium,Non-profit,Full coverage,About the same,Offered,Offered,Allowed,50.0,40.0,60.0,60.0,40.0,30.0,0.0,28.0,Balanced
5,Hospital Worksites,Small,Non-profit,Full coverage,About the same,Offered,Offered,Allowed,20.0,25.0,65.0,80.0,25.0,5.0,0.0,15.0,Balanced
6,Hospital Worksites,Large,Non-profit,Partial coverage,Larger,Offered,Offered,Allowed,20.0,20.0,80.0,70.0,60.0,5.0,10.0,NaN,Mostly women
7,Hospital Worksites,Large,"For profit, public",Partial coverage,About the same,Not offered,Offered,Not allowed,27.0,14.0,75.0,60.0,50.0,5.0,0.0,15.0,Mostly women
8,Hospital Worksites,Large,NaN,Full coverage,Larger,Offered,Offered,Allowed,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN
9,Hospital Worksites,Large,Non-profit,Partial coverage,About the same,Offered,Offered,Allowed,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Part h
Currently all of the categorical features in the data are stored as "object" data types, which is the same data type as strings. But there are additional methods and display properties that are available if we instead convert these features to a "category" data type.

Change the data type of all categorical features in the working data from "object" to "category". [4 points]

In [286]:
cat_columns = [c for c in working_data.columns if not c.startswith('Percent')]
working_data[cat_columns] = working_data[cat_columns].astype('category')

In [287]:
working_data.dtypes

Industry                              category
Size                                  category
Org_Category                          category
Employer_Payment_Offer                category
Proportion_Employee_Payment_Change    category
Part_Time_Health_Insurance            category
Health_Education_Programs             category
Work_From_Home                        category
Percent_Age_LT_30                      float64
Percent_Age_GTE_60                     float64
Percent_Female                         float64
Percent_Hourly_NonExempt               float64
Percent_NonDaytime_Shift               float64
Percent_Remote                         float64
Percent_Unionized                      float64
Percent_Annual_Turnover                float64
Gender_Balance                        category
dtype: object

## Problem 3
Just like SQL is a mechanism for deriving answers from the information stored within a database, pandas can be a tool for getting answers to questions using a dataframe.

The following questions ask you to perform some simple analyses using data wrangling alone.

### Part a
I am looking for a job with a small workplace that lets me work from home, but I am worried about whether the job will provide me with health insurance.

Filter the data to only those rows that represent small workplaces that allow employees to work from home. Then report how many of these workplaces offer full insurance, partial insurance, and no insurance. Report the percent, cumulative count, and cumulative percent in addition to the counts. 

This is much easier if you use the `sidetable` package. Any answer that does not use `sidetable` will receive no credit.
[4 points]

In [288]:
working_data.query('Size=="Small" & Work_From_Home=="Allowed"').stb.freq(['Employer_Payment_Offer'])

/Users/ceceliaauerswald/miniconda3/envs/ds6001fall2025/lib/python3.13/site-packages/sidetable/sidetable.py:103: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  group_data = self._obj.groupby(cols).size().reset_index(


,Employer_Payment_Offer,count,percent,cumulative_count,cumulative_percent
0,Full coverage,324,46.285714,324,46.285714
1,Partial coverage,310,44.285714,634,90.571429
2,No coverage,66,9.428571,700,100.000000


### Part b
Anything that can be done in SQL can be done with `pandas`. The next several questions ask you to write `pandas` code to match a given SQL query. 

But first, to check that the SQL query and `pandas` code yield the same result, create a new database using the `sqlite3` package and input the cleaned WHA data as a table in this database. [4 points]

In [289]:
work_db = sqlite3.connect("workplaces.db") #create local db
working_data.to_sql("workplaces", con=work_db, index=False, chunksize=1000, if_exists='replace') #load in data to db

2842

In [290]:
#check data is loaded into db
myquery = '''
SELECT *
FROM workplaces
'''
db_data = pd.read_sql_query(myquery, work_db)
db_data.count()==working_data.count()

Industry                              True
Size                                  True
Org_Category                          True
Employer_Payment_Offer                True
Proportion_Employee_Payment_Change    True
Part_Time_Health_Insurance            True
Health_Education_Programs             True
Work_From_Home                        True
Percent_Age_LT_30                     True
Percent_Age_GTE_60                    True
Percent_Female                        True
Percent_Hourly_NonExempt              True
Percent_NonDaytime_Shift              True
Percent_Remote                        True
Percent_Unionized                     True
Percent_Annual_Turnover               True
Gender_Balance                        True
dtype: bool

### Part c
Some hospital workplaces lowered their insurance premiums. I want to know whether these hospitals were large, medium, or small employers, for profit or non-profit, whether they offer full or partial insurance, and whether they have part-time employees. Using SQL, I can get an answer to these questions with this query:
```
SELECT size, 
    type, 
    premiums AS insurance, 
    parttime
FROM whpps
WHERE industry = 'Hospitals' 
    AND premium_change='Smaller'
```
Write pandas code that replicates the results of this SQL query. Your feature names might be different from the ones listed in the query, depending on the column names you chose earlier. [8 points]

In [291]:
my_sql_query = '''
SELECT Size, 
    Org_Category, 
    Employer_Payment_Offer as Insurance, 
    Part_Time_Health_Insurance
FROM workplaces
WHERE Industry = "Hospital Worksites"
    AND Proportion_Employee_Payment_Change="Smaller"
'''
sql_results = pd.read_sql_query(my_sql_query, work_db)
sql_results

,Size,Org_Category,Insurance,Part_Time_Health_Insurance
0,Medium,"For profit, private",Full coverage,Offered
1,Medium,None,Partial coverage,Offered
2,Medium,Non-profit,Partial coverage,Offered
3,Medium,Non-profit,Full coverage,Offered
4,Medium,Non-profit,Full coverage,None
5,Large,Non-profit,Partial coverage,Offered
6,Large,Non-profit,Partial coverage,Offered
7,Medium,Non-profit,Partial coverage,Offered
8,Large,Non-profit,Partial coverage,Offered
9,Small,Non-profit,Full coverage,Offered


In [292]:
pandas_results = db_data.query('Industry=="Hospital Worksites" & Proportion_Employee_Payment_Change=="Smaller"')[['Size','Org_Category', 'Employer_Payment_Offer', 'Part_Time_Health_Insurance']]
pandas_results.rename({'Employer_Payment_Offer': 'Insurance'}, axis='columns', inplace=True)
pandas_results

,Size,Org_Category,Insurance,Part_Time_Health_Insurance
3,Medium,"For profit, private",Full coverage,Offered
11,Medium,None,Partial coverage,Offered
48,Medium,Non-profit,Partial coverage,Offered
51,Medium,Non-profit,Full coverage,Offered
75,Medium,Non-profit,Full coverage,None
97,Large,Non-profit,Partial coverage,Offered
187,Large,Non-profit,Partial coverage,Offered
191,Medium,Non-profit,Partial coverage,Offered
214,Large,Non-profit,Partial coverage,Offered
229,Small,Non-profit,Full coverage,Offered


### Part d
Are there differences across industry in the percent of the workforce that are women, under 30, or over 60?

Using SQL, I can get an answer to these questions with this query:
```
SELECT industry, 
    AVG(percent_female) AS percent_female, 
    AVG(percent_under30) AS percent_under30,
    AVG(percent_over60) AS percent_over60
FROM whpps
GROUP BY industry
ORDER BY percent_female DESC
```
Write pandas code that replicates the results of this SQL query. Your feature names might be different from the ones listed in the query, depending on the column names you chose earlier. [8 points]

In [293]:
my_sql_query = '''
SELECT Industry, 
    AVG(Percent_Female) AS Percent_Female, 
    AVG(Percent_Age_LT_30) AS Percent_Age_LT_30,
    AVG(Percent_Age_GTE_60) AS Percent_Age_GTE_60
FROM workplaces
GROUP BY Industry
ORDER BY Percent_Female DESC
'''
sql_results = pd.read_sql_query(my_sql_query, work_db)
sql_results

,Industry,Percent_Female,Percent_Age_LT_30,Percent_Age_GTE_60
0,"Education, Health, and Social Assistance",80.657143,25.745665,11.349570
1,Hospital Worksites,76.427027,27.213793,16.489655
2,Entertainment and Non-public Services,53.804416,38.566343,11.544872
3,"Info, Business, Tech, and Property",50.632184,23.821752,12.465465
4,Public Administration,39.056738,21.015625,15.015385
5,Trade and Transportation,32.657258,29.108696,12.584034
6,Food and Resource Production,20.328605,22.257143,10.690355


In [294]:
pandas_results = db_data.groupby('Industry').agg({
    'Percent_Female': 'mean',
    'Percent_Age_LT_30': 'mean',
    'Percent_Age_GTE_60': 'mean'
}).sort_values(by='Percent_Female', ascending=False)
pandas_results

,Percent_Female,Percent_Age_LT_30,Percent_Age_GTE_60
Industry,,,
"Education, Health, and Social Assistance",80.657143,25.745665,11.349570
Hospital Worksites,76.427027,27.213793,16.489655
Entertainment and Non-public Services,53.804416,38.566343,11.544872
"Info, Business, Tech, and Property",50.632184,23.821752,12.465465
Public Administration,39.056738,21.015625,15.015385
Trade and Transportation,32.657258,29.108696,12.584034
Food and Resource Production,20.328605,22.257143,10.690355


### Part e
Are there differences in whether full, partial, or no health insurance is provided to employees based on whether the workplace is mostly men, mostly women, or is gender balanced?

Using SQL, I can get an answer to this question with this query:
```
SELECT gender_balance, 
    premiums, 
    COUNT(*)
FROM whpps
GROUP BY gender_balance, premiums
HAVING gender_balance is NOT NULL 
    AND premiums is NOT NULL
```
Write pandas code that replicates the results of this SQL query. Your feature names might be different from the ones listed in the query, depending on the column names you chose earlier. 

Note: the rows of your pandas output will be sorted in a different order than the rows of the SQL query. That happens due to the default sorting order (SQL sorts alphabetically, pandas sorts according to the order supplied in the `pd.cut()` method you used earlier.) It's a pain to get the sorting to match, so don't! Also, don't worry if the columns have different names. Just confirm the same data is contained in each result. [8 points]

In [295]:
my_sql_query = '''
SELECT Gender_Balance, 
    Employer_Payment_Offer, 
    COUNT(*)
FROM workplaces
GROUP BY Gender_Balance, Employer_Payment_Offer
HAVING Gender_Balance is NOT NULL
    AND Employer_Payment_Offer is NOT NULL
'''
sql_results = pd.read_sql_query(my_sql_query, work_db)
sql_results

,Gender_Balance,Employer_Payment_Offer,COUNT(*)
0,Balanced,Full coverage,226
1,Balanced,No coverage,77
2,Balanced,Partial coverage,271
3,Mostly men,Full coverage,301
4,Mostly men,No coverage,91
5,Mostly men,Partial coverage,332
6,Mostly women,Full coverage,267
7,Mostly women,No coverage,107
8,Mostly women,Partial coverage,333


In [297]:
pandas_results = db_data[db_data.Gender_Balance.notnull() & db_data.Employer_Payment_Offer.notnull()].groupby(['Gender_Balance', 'Employer_Payment_Offer']).size()
pd.DataFrame(pandas_results)

0
Gender_Balance Employer_Payment_Offer     
Balanced       Full coverage           226
               No coverage              77
               Partial coverage        271
Mostly men     Full coverage           301
               No coverage              91
               Partial coverage        332
Mostly women   Full coverage           267
               No coverage             107
               Partial coverage        333